From 975a39ea719eeae82918b0fb034f9026dbe55a8a Mon Sep 17 00:00:00 2001
From: Christoph Reiter <reiter.christoph@gmail.com>
Date: Tue, 30 Dec 2025 18:07:22 +0100
Subject: [PATCH] content type to mime type conversion (v2)

The code was mixing content types with mime types in various places.
Make sure to convert content types to mime types where mime types
are needed.

In eog-window.c a mime type was used instead of a content
type instead, move the unref after the content type is no longer needed.

Also fix the naming of assigned variables for all
eog_util_get_content_type_with_fallback() call sites, even if they are
correct, to make things clearer.

This is similar to https://gitlab.gnome.org/GNOME/eog/-/merge_requests/112
except I added a eog_util_get_mime_type_with_fallback() helper similar to
eog_util_get_content_type_with_fallback() + some missing g_free calls, plus
this is on top of current master, which changed a lot since then.

This fixes showing images in directories on Windows for example.

Fixes #214
---
 src/eog-file-chooser.c     |  3 +--
 src/eog-image.c            |  2 +-
 src/eog-jobs.c             | 10 ++++++----
 src/eog-list-store.c       | 20 +++++++++++++-------
 src/eog-metadata-sidebar.c |  6 +++---
 src/eog-remote-presenter.c |  6 +++---
 src/eog-thumb-view.c       |  8 ++++----
 src/eog-thumbnail.c        |  2 +-
 src/eog-util.c             | 13 +++++++++++++
 src/eog-util.h             |  3 +++
 src/eog-window.c           | 12 ++++++------
 11 files changed, 54 insertions(+), 31 deletions(-)

diff --git a/src/eog-file-chooser.c b/src/eog-file-chooser.c
index ff6c9f7b..8f071b4a 100644
--- a/src/eog-file-chooser.c
+++ b/src/eog-file-chooser.c
@@ -347,8 +347,7 @@ update_preview_cb (GtkFileChooser *file_chooser, gpointer data)
 		} else if (g_file_info_get_size (file_info) <= 100000) {
 			/* read files smaller than 100kb directly */
 
-			gchar *mime_type = g_content_type_get_mime_type (
-						eog_util_get_content_type_with_fallback (file_info));
+			gchar *mime_type = eog_util_get_mime_type_with_fallback (file_info);
 
 
 			if (G_LIKELY (mime_type)) {
diff --git a/src/eog-image.c b/src/eog-image.c
index f52f55d1..4e61f87e 100644
--- a/src/eog-image.c
+++ b/src/eog-image.c
@@ -628,7 +628,7 @@ eog_image_get_file_info (EogImage *img,
 			*bytes = g_file_info_get_size (file_info);
 
 		if (mime_type) {
-			*mime_type = g_strdup (eog_util_get_content_type_with_fallback (file_info));
+			*mime_type = eog_util_get_mime_type_with_fallback (file_info);
 		}
 		g_object_unref (file_info);
 	}
diff --git a/src/eog-jobs.c b/src/eog-jobs.c
index f5609226..465715ed 100644
--- a/src/eog-jobs.c
+++ b/src/eog-jobs.c
@@ -768,14 +768,16 @@ filter_files (GSList *files, GList **file_list, GList **error_list)
 				/* Workaround for gvfs backends that
 				   don't set the GFileType. */
 				if (G_UNLIKELY (type == G_FILE_TYPE_UNKNOWN)) {
-					const gchar *ctype;
+					gchar *mime_type;
 
-					ctype = eog_util_get_content_type_with_fallback (file_info);
+					mime_type = eog_util_get_mime_type_with_fallback (file_info);
 
-					/* If the content type is supported
+					/* If the mime type is supported
 					   adjust the file_type */
-					if (eog_image_is_supported_mime_type (ctype))
+					if (eog_image_is_supported_mime_type (mime_type))
 						type = G_FILE_TYPE_REGULAR;
+
+					g_free (mime_type);
 				}
 
 				g_object_unref (file_info);
diff --git a/src/eog-list-store.c b/src/eog-list-store.c
index 27c8864a..a2fd116c 100644
--- a/src/eog-list-store.c
+++ b/src/eog-list-store.c
@@ -456,7 +456,7 @@ file_monitor_changed_cb (GFileMonitor *monitor,
 			 GFileMonitorEvent event,
 			 EogListStore *store)
 {
-	const char *mimetype;
+	char *mimetype;
 	GFileInfo *file_info;
 	GtkTreeIter iter;
 	EogImage *image;
@@ -472,7 +472,7 @@ file_monitor_changed_cb (GFileMonitor *monitor,
 		if (file_info == NULL) {
 			break;
 		}
-		mimetype = eog_util_get_content_type_with_fallback (file_info);
+		mimetype = eog_util_get_mime_type_with_fallback (file_info);
 
 		if (is_file_in_list_store_file (store, file, &iter)) {
 			if (eog_image_is_supported_mime_type (mimetype)) {
@@ -493,6 +493,7 @@ file_monitor_changed_cb (GFileMonitor *monitor,
 				eog_list_store_append_image_from_file (store, file, caption);
 			}
 		}
+		g_free (mimetype);
 		g_object_unref (file_info);
 		break;
 	case G_FILE_MONITOR_EVENT_PRE_UNMOUNT:
@@ -525,7 +526,7 @@ file_monitor_changed_cb (GFileMonitor *monitor,
 			if (file_info == NULL) {
 				break;
 			}
-			mimetype = eog_util_get_content_type_with_fallback (file_info);
+			mimetype = eog_util_get_mime_type_with_fallback (file_info);
 
 			if (eog_image_is_supported_mime_type (mimetype)) {
 				const gchar *caption;
@@ -533,6 +534,7 @@ file_monitor_changed_cb (GFileMonitor *monitor,
 				caption = g_file_info_get_display_name (file_info);
 				eog_list_store_append_image_from_file (store, file, caption);
 			}
+			g_free (mimetype);
 			g_object_unref (file_info);
 		}
 		break;
@@ -544,11 +546,12 @@ file_monitor_changed_cb (GFileMonitor *monitor,
 		if (file_info == NULL) {
 			break;
 		}
-		mimetype = eog_util_get_content_type_with_fallback (file_info);
+		mimetype = eog_util_get_mime_type_with_fallback (file_info);
 		if (is_file_in_list_store_file (store, file, &iter) &&
 		    eog_image_is_supported_mime_type (mimetype)) {
 			eog_list_store_thumbnail_refresh (store, &iter);
 		}
+		g_free (mimetype);
 		g_object_unref (file_info);
 		break;
 	case G_FILE_MONITOR_EVENT_RENAMED:
@@ -560,7 +563,7 @@ file_monitor_changed_cb (GFileMonitor *monitor,
 		if (file_info == NULL) {
 			break;
 		}
-		mimetype = eog_util_get_content_type_with_fallback (file_info);
+		mimetype = eog_util_get_mime_type_with_fallback (file_info);
 
 		if (is_file_in_list_store_file (store, other_file, &iter)) {
 			gtk_tree_model_get (GTK_TREE_MODEL (store), &iter,
@@ -579,6 +582,7 @@ file_monitor_changed_cb (GFileMonitor *monitor,
 			eog_list_store_remove (store, &iter);
 		}
 
+		g_free (mimetype);
 		g_object_unref (file_info);
 		break;
 	case G_FILE_MONITOR_EVENT_CHANGED:
@@ -599,9 +603,10 @@ directory_visit (GFile *directory,
 {
 	GFile *child;
 	gboolean load_uri = FALSE;
-	const char *mime_type, *name;
+	const char *name;
+	char *mime_type;
 
-	mime_type = eog_util_get_content_type_with_fallback (children_info);
+	mime_type = eog_util_get_mime_type_with_fallback (children_info);
 	name = g_file_info_get_name (children_info);
 
         if (!g_str_has_prefix (name, ".")) {
@@ -609,6 +614,7 @@ directory_visit (GFile *directory,
 			load_uri = TRUE;
 		}
 	}
+	g_free (mime_type);
 
 	if (load_uri) {
 		const gchar *caption;
diff --git a/src/eog-metadata-sidebar.c b/src/eog-metadata-sidebar.c
index 06a702b5..36f84560 100644
--- a/src/eog-metadata-sidebar.c
+++ b/src/eog-metadata-sidebar.c
@@ -158,10 +158,10 @@ eog_metadata_sidebar_update_general_section (EogMetadataSidebar *sidebar)
 	if (file_info == NULL) {
 		str = g_strdup (_("Unknown"));
 	} else {
-		const gchar *mime_str;
+		const gchar *content_type;
 
-		mime_str = eog_util_get_content_type_with_fallback (file_info);
-		str = g_content_type_get_description (mime_str);
+		content_type = eog_util_get_content_type_with_fallback (file_info);
+		str = g_content_type_get_description (content_type);
 		g_object_unref (file_info);
 	}
 	gtk_label_set_text (GTK_LABEL (priv->type_label), str);
diff --git a/src/eog-remote-presenter.c b/src/eog-remote-presenter.c
index fe75a5bb..a744a6e6 100644
--- a/src/eog-remote-presenter.c
+++ b/src/eog-remote-presenter.c
@@ -327,7 +327,7 @@ eog_remote_presenter_update (EogRemotePresenter *remote_presenter,
         gchar *size_str;
         GFile *file, *parent_file;
         GFileInfo *file_info;
-        const char *mime_str;
+        const char *content_type;
         char *type_str;
         gint width, height;
         goffset bytes;
@@ -357,8 +357,8 @@ eog_remote_presenter_update (EogRemotePresenter *remote_presenter,
         if (file_info == NULL) {
                 type_str = g_strdup (_("Unknown"));
         } else {
-                mime_str = eog_util_get_content_type_with_fallback (file_info);
-                type_str = g_content_type_get_description (mime_str);
+                content_type = eog_util_get_content_type_with_fallback (file_info);
+                type_str = g_content_type_get_description (content_type);
                 g_object_unref (file_info);
         }
 
diff --git a/src/eog-thumb-view.c b/src/eog-thumb-view.c
index 36286954..8030a9ed 100644
--- a/src/eog-thumb-view.c
+++ b/src/eog-thumb-view.c
@@ -487,7 +487,7 @@ thumbview_get_tooltip_string (EogImage *image)
 	gint width, height;
 	GFile *file;
 	GFileInfo *file_info;
-	const char *mime_str;
+	const char *content_type;
 	gchar *tooltip_string;
 #ifdef HAVE_EXIF
 	ExifData *exif_data;
@@ -508,15 +508,15 @@ thumbview_get_tooltip_string (EogImage *image)
 		return NULL;
 	}
 
-	mime_str = eog_util_get_content_type_with_fallback (file_info);
+	content_type = eog_util_get_content_type_with_fallback (file_info);
 
-	if (G_UNLIKELY (mime_str == NULL)) {
+	if (G_UNLIKELY (content_type == NULL)) {
 		g_free (bytes);
 		g_object_unref (image);
 		return NULL;
 	}
 
-	type_str = g_content_type_get_description (mime_str);
+	type_str = g_content_type_get_description (content_type);
 	g_object_unref (file_info);
 
 	if (width > -1 && height > -1) {
diff --git a/src/eog-thumbnail.c b/src/eog-thumbnail.c
index 64c73e7a..66043b5e 100644
--- a/src/eog-thumbnail.c
+++ b/src/eog-thumbnail.c
@@ -169,7 +169,7 @@ eog_thumb_data_new (GFile *file, GError **error)
 		/* if available, copy data */
 		data->mtime = g_file_info_get_attribute_uint64 (file_info,
 								G_FILE_ATTRIBUTE_TIME_MODIFIED);
-		data->mime_type = g_strdup (eog_util_get_content_type_with_fallback (file_info));
+		data->mime_type = eog_util_get_mime_type_with_fallback (file_info);
 
 		data->failed_thumb_exists = g_file_info_get_attribute_boolean (file_info,
 									       G_FILE_ATTRIBUTE_THUMBNAILING_FAILED);
diff --git a/src/eog-util.c b/src/eog-util.c
index fab4c5e4..ba51a89f 100644
--- a/src/eog-util.c
+++ b/src/eog-util.c
@@ -493,6 +493,19 @@ eog_util_get_content_type_with_fallback (GFileInfo *file_info)
 	return NULL;
 }
 
+char*
+eog_util_get_mime_type_with_fallback (GFileInfo *file_info)
+{
+	g_return_val_if_fail (file_info != NULL, NULL);
+
+	const char* content_type = eog_util_get_content_type_with_fallback (file_info);
+	if (content_type != NULL) {
+		return g_content_type_get_mime_type (content_type);
+	}
+
+	return NULL;
+}
+
 
 /* Portal */
 
diff --git a/src/eog-util.h b/src/eog-util.h
index f96ee5fc..37c1c196 100644
--- a/src/eog-util.h
+++ b/src/eog-util.h
@@ -71,6 +71,9 @@ gchar *eog_util_create_width_height_string (gint width, gint height);
 G_GNUC_INTERNAL
 const     char *eog_util_get_content_type_with_fallback	(GFileInfo *file_info);
 
+G_GNUC_INTERNAL
+char* eog_util_get_mime_type_with_fallback (GFileInfo *file_info);
+
 #ifdef HAVE_LIBPORTAL
 G_GNUC_INTERNAL
 gboolean eog_util_is_running_inside_flatpak  (void);
diff --git a/src/eog-window.c b/src/eog-window.c
index c8697270..e778f34a 100644
--- a/src/eog-window.c
+++ b/src/eog-window.c
@@ -852,7 +852,7 @@ add_file_to_recent_files (GFile *file)
 	recent_data = g_slice_new (GtkRecentData);
 	recent_data->display_name = NULL;
 	recent_data->description = NULL;
-	recent_data->mime_type = (gchar *) eog_util_get_content_type_with_fallback (file_info);
+	recent_data->mime_type = eog_util_get_mime_type_with_fallback (file_info);
 	recent_data->app_name = EOG_RECENT_FILES_APP_NAME;
 	recent_data->app_exec = g_strdup ("eog %u");
 	recent_data->groups = groups;
@@ -863,6 +863,7 @@ add_file_to_recent_files (GFile *file)
 	                             recent_data);
 
 	g_free (recent_data->app_exec);
+	g_free (recent_data->mime_type);
 	g_free (text_uri);
 	g_object_unref (file_info);
 
@@ -1086,28 +1087,27 @@ eog_window_open_file_chooser_dialog (EogWindow *window)
 	GtkWidget *dialog;
 	GFileInfo *file_info;
 	GFile *file;
-	const gchar *mime_type = NULL;
+	const gchar *content_type;
 
 	file = eog_image_get_file (window->priv->image);
 	file_info = g_file_query_info (file,
 				       G_FILE_ATTRIBUTE_STANDARD_CONTENT_TYPE","
 				       G_FILE_ATTRIBUTE_STANDARD_FAST_CONTENT_TYPE,
 				       0, NULL, NULL);
-	mime_type = g_content_type_get_mime_type (
-			eog_util_get_content_type_with_fallback (file_info));
-	g_object_unref (file_info);
+	content_type = eog_util_get_content_type_with_fallback (file_info);
 
 	dialog = gtk_app_chooser_dialog_new_for_content_type (GTK_WINDOW (window),
 							      GTK_DIALOG_MODAL |
 							      GTK_DIALOG_DESTROY_WITH_PARENT |
 							      GTK_DIALOG_USE_HEADER_BAR,
-							      mime_type);
+							      content_type);
 	gtk_widget_show (dialog);
 
 	g_signal_connect_object (dialog, "response",
 				 G_CALLBACK (app_chooser_dialog_response_cb),
 				 window, 0);
 
+	g_object_unref (file_info);
 	g_object_unref (file);
 }
 
-- 
2.52.0

